webapp-django/crashstats/cron/management/commands/cron.py

```diff
@@ -0,0 +1,464 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import contextlib
+import datetime
+import logging
+import re
+import sys
+import time
+import traceback
+
+import markus
+
+from django.core.management import call_command
+from django.core.management.base import BaseCommand
+from django.utils import timezone
+
+from crashstats.cron.models import Job, Log
+
+
+logger = logging.getLogger('crashstats.crontabber')
+
+metrics = markus.get_metrics('crontabber')
+
+
+# NOTE(willkg): Times are in UTC. These are Django-based jobs. The other
+# crontabber handles the jobs that haven't been converted, yet.
+#
+# NOTE(willkg): You can't have the same job in here twice with two different
+# frequencies. If we ever need to do that, we'll need to stop using "cmd" as
+# the key in the db.
+3
+# cmd, frequency, time, backfill
+JOBS = [
+    {
+        # Test cron job
+        'cmd': 'crontest',
+        'frequency': '1d',
+    },
+]
+
+# Map of cmd -> job_spec
+JOBS_MAP = dict([(job_spec['cmd'], job_spec) for job_spec in JOBS])
+
+
+# The maximum time we let a job go for before we declare it a zombie (in seconds)
+MAX_ONGOING = 60 * 60 * 2
+
+
+# Error retry time (in seconds)
+ERROR_RETRY_TIME = 60
+
+
+class FrequencyDefinitionError(Exception):
+    pass
+
+
+class JobNotFoundError(Exception):
+    pass
+
+
+class TimeDefinitionError(Exception):
+    pass
+
+
+class OngoingJobError(Exception):
+    """Error raised when a job is currently running."""
+
+    pass
+
+
+FREQUENCY_RE = re.compile(r'^(\d+)([^\d])$')
+
+
+def convert_frequency(value):
+    """return the number of seconds that a certain frequency string represents.
+    For example: `1d` means 1 day which means 60 * 60 * 24 seconds.
+    The recognized formats are:
+        10d  : 10 days
+        3m   : 3 minutes
+        12h  : 12 hours
+    """
+    match = FREQUENCY_RE.match(value)
+    if not match:
+        raise FrequencyDefinitionError(value)
+
+    number = int(match.group(1))
+    unit = match.group(2)
+
+    if unit == 'h':
+        number *= 60 * 60
+    elif unit == 'm':
+        number *= 60
+    elif unit == 'd':
+        number *= 60 * 60 * 24
+    elif unit:
+        raise FrequencyDefinitionError(value)
+    return number
+
+
+VALID_TIME_RE = re.compile(r'^\d\d:\d\d$')
+
+
+def check_time(value):
+    """Validate it's an appropriate time in HH:MM form."""
+    if not VALID_TIME_RE.match(value):
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+    hh, mm = value.split(':')
+    if int(hh) > 23 or int(mm) > 59:
+        raise TimeDefinitionError("Invalid definition of time %r" % value)
+
+
+def get_matching_job_specs(cmds):
+    """Return list of matching job specs for this cmd.
+
+    :arg cmds: "all" returns list of all job_specs, cmds return single job_spec,
+        list of cmds returns list of job_specs
+
+    """
+    if cmds == ['all']:
+        return JOBS
+    if isinstance(cmds, str):
+        return JOBS_MAP.get(cmds)
```

Comment:
I just double-checked the code and I don't think this can ever happen--it's always a list of cmds. Should fix this and the docstring.